packages/hierarchies-react/src/presentation-hierarchies-react/internal/TreeLoader.ts

```diff
@@ -61,56 +67,54 @@ export class TreeLoader implements ITreeLoader {
-        const nodeProps = {
-          id: `${infoNodeIdBase}-Unknown`,
-          parentId: parent.id,
-        };
         let hierarchyLoadErrorType: "unknown" | "timeout" = "unknown";
         if (err instanceof Error) {
-          nodeProps.id = `${infoNodeIdBase}-${err.message}`;
           if (isRowsLimitError(err)) {
             this._onHierarchyLimitExceeded({ parentId: parent.id, filter: instanceFilter, limit: err.limit });
-            return of([{ ...nodeProps, type: "ResultSetTooLarge" as const, resultSetSizeLimit: err.limit }]);
+            return of({ id: `${infoNodeIdBase}-${err.message}`, type: "ResultSetTooLarge" as const, resultSetSizeLimit: err.limit });
           }
           if (isTimeoutError(err)) {
             hierarchyLoadErrorType = "timeout";
           }
         }
 
         this._onHierarchyLoadError({ parentId: parent.id, type: hierarchyLoadErrorType, error: err });
-        return of([
-          {
-            ...nodeProps,
-            type: "Unknown" as const,
-            message: "Failed to create hierarchy level",
-          },
-        ]);
+        return of({
+          id: `${infoNodeIdBase}-Unknown`,
+          type: "Unknown" as const,
+          message: "Failed to create hierarchy level",
+        });
       }),
-      map(
-        (childNodes): LoadedTreePart => ({
+      map((childNodes): LoadedTreePart => {
+        if ("length" in childNodes) {
+          return instanceFilter && childNodes.length === 0
+            ? {
+                parentId: parent.id,
+                error: {
+                  id: `${infoNodeIdBase}-no-filter-matches`,
+                  type: "NoFilterMatches" as const,
+                },
+              }
+            : {
+                parentId: parent.id,
+                loadedNodes: childNodes.map(treeModelNodesFactory),
+              };
+        }
+        return {
           parentId: parent.id,
-          loadedNodes:
-            instanceFilter && childNodes.length === 0
-              ? [
-                  {
-                    id: `${infoNodeIdBase}-no-filter-matches`,
-                    parentId: parent.id,
-                    type: "NoFilterMatches" as const,
-                  },
-                ]
-              : childNodes.map(treeModelNodesFactory),
-        }),
-      ),
+          error: childNodes,
+        };
+      }),
     );
   }
 
   public loadNodes({ shouldLoadChildren, ...options }: LoadNodesOptions) {
     return this.loadChildren(options).pipe(
-      expand((loadedPart) =>
-        from(loadedPart.loadedNodes.filter((node): node is TreeModelHierarchyNode => isTreeModelHierarchyNode(node) && shouldLoadChildren(node))).pipe(
+      expand((loadedPart) => {
+        return from("loadedNodes" in loadedPart ? loadedPart.loadedNodes.filter((node) => shouldLoadChildren(node)) : []).pipe(
           mergeMap((node) => this.loadChildren({ ...options, parent: node })),
-        ),
-      ),
+        );
+      }),
```

Comment:
Could we also move the `filter((node) => shouldLoadChildren(node))` part into the `pipe` (use rxjs filter)?


```suggestion
      expand((loadedPart) => 
        "loadedNodes" in loadedPart
          ? from(loadedPart.loadedNodes.filter((node) => shouldLoadChildren(node))).pipe(
              mergeMap((node) => this.loadChildren({ ...options, parent: node })),
            )
          : EMPTY
      ),
```

Reply:
Added suggestion, also changed filter to rxjs.